cmd/aws-vpc-cni/main.go

```diff
@@ -0,0 +1,365 @@
+	envEnPrefixDelegation    = "ENABLE_PREFIX_DELEGATION"
+	envWarmIPTarget          = "WARM_IP_TARGET"
+	envMinIPTarget           = "MINIMUM_IP_TARGET"
+	envWarmPrefixTarget      = "WARM_PREFIX_TARGET"
+	envEnBandwidthPlugin     = "ENABLE_BANDWIDTH_PLUGIN"
+	envEnIPv6                = "ENABLE_IPv6"
+)
+
+func getEnv(env, def string) string {
+	if val, ok := os.LookupEnv(env); ok {
+		return val
+	}
+	return def
+}
+
+// zero values of structs with omitempty is not supported, Golang issue for - https://github.com/golang/go/issues/11939
+// cni/types DNS struct is missing omitstring
+// NetConfList describes an ordered list of networks.
+type NetConfList struct {
+	CNIVersion string `json:"cniVersion,omitempty"`
+
+	Name         string     `json:"name,omitempty"`
+	DisableCheck bool       `json:"disableCheck,omitempty"`
+	Plugins      []*NetConf `json:"plugins,omitempty"`
+}
+
+// NetConf stores the common network config for the CNI plugin
+type NetConf struct {
+	CNIVersion string `json:"cniVersion,omitempty"`
+
+	Name         string          `json:"name,omitempty"`
+	Type         string          `json:"type,omitempty"`
+	Capabilities map[string]bool `json:"capabilities,omitempty"`
+	IPAM         *IPAMConfig     `json:"ipam,omitempty"`
+	DNS          *types.DNS      `json:"dns,omitempty"`
+
+	RawPrevResult map[string]interface{} `json:"prevResult,omitempty"`
+	PrevResult    types.Result           `json:"-"`
+
+	// Interface inside container to create
+	IfName string `json:"ifName,omitempty"`
+
+	Enabled string `json:"enabled,,omitempty"`
+
+	// IP to use as SNAT target
+	NodeIP net.IP `json:"nodeIP,omitempty"`
+
+	VethPrefix string `json:"vethPrefix,omitempty"`
+
+	// MTU for eth0
+	MTU string `json:"mtu,omitempty"`
+
+	PluginLogFile string `json:"pluginLogFile,omitempty"`
+
+	PluginLogLevel string `json:"pluginLogLevel,omitempty"`
+}
+
+//Ref : https://github.com/containernetworking/plugins/plugins/ipam/host-local/backend/allocator/config.go
+type IPAMConfig struct {
+	*Range
+	Name       string         `json:"name,omitempty"`
+	Type       string         `json:"type,omitempty"`
+	Routes     []*types.Route `json:"routes,omitempty"`
+	DataDir    string         `json:"dataDir,omitempty"`
+	ResolvConf string         `json:"resolvConf,omitempty"`
+	Ranges     []RangeSet     `json:"ranges"`
+	IPArgs     []net.IP       `json:"-"` // Requested IPs from CNI_ARGS and args
+}
+
+type RangeSet []Range
+
+type Range struct {
+	RangeStart net.IP      `json:"rangeStart,omitempty"` // The first ip, inclusive
+	RangeEnd   net.IP      `json:"rangeEnd,omitempty"`   // The last ip, inclusive
+	Subnet     types.IPNet `json:"subnet"`
+	Gateway    net.IP      `json:"gateway,omitempty"`
+}
+
+func waitForIPAM() bool {
+	for {
+		cmd := exec.Command("./grpc-health-probe", "-addr", "127.0.0.1:50051", ">", "/dev/null", "2>&1")
+		var outb bytes.Buffer
+		cmd.Stdout = &outb
+		cmd.Run()
+		if outb.String() == "" {
+			return true
+		}
+		time.Sleep(1 * time.Second)
```

Comment:
nit: do we need a sleep here? Won't `cmd` already have a built-in grpc timeout that should be sufficient?

Reply:
Yeah we can remove it, this was based of the old code and it was removed recently.

https://github.com/aws/amazon-vpc-cni-k8s/blob/v1.10.3/scripts/entrypoint.sh#L120